cli/api/commands/compile.ts

```diff
@@ -14,22 +17,53 @@ export async function compile(
   compileConfig: dataform.ICompileConfig = {}
 ): Promise<dataform.CompiledGraph> {
   let compiledGraph = dataform.CompiledGraph.create();
-  let workflowSettingsDataformCoreVersion = "";
-  try {
-    workflowSettingsDataformCoreVersion = await runInstallIfWorkflowSettingsDataformCoreVersion(
-      compileConfig.projectDir
-    );
 
-    const result = await CompileChildProcess.forkProcess().compile(compileConfig);
+  const resolvedProjectPath = path.resolve(compileConfig.projectDir);
+  const packageJsonPath = path.join(resolvedProjectPath, "package.json");
+  const projectNodeModulesPath = path.join(resolvedProjectPath, "node_modules");
+  const temporaryInstallPath = tmp.dirSync().name;
 
-    const decodedResult = decode64(dataform.CoreExecutionResponse, result);
-    compiledGraph = dataform.CompiledGraph.create(decodedResult.compile.compiledGraph);
+  const workflowSettingsDataformCoreVersion = readDataformCoreVersionFromWorkflowSettings(
+    resolvedProjectPath
+  );
 
-    compiledGraph.tables.forEach(setOrValidateTableEnumType);
-  } finally {
-    if (workflowSettingsDataformCoreVersion) {
-      cleanupNpmFiles(compileConfig.projectDir);
+  if (!workflowSettingsDataformCoreVersion && !fs.existsSync(packageJsonPath)) {
+    throw new Error(MISSING_CORE_VERSION_ERROR);
+  }
+
+  // Packages are installed to a temporary directory in order to minimize interfering with user's
+  // project directories.
+  if (workflowSettingsDataformCoreVersion) {
+    if (fs.existsSync(projectNodeModulesPath)) {
+      fs.rmdirSync(projectNodeModulesPath);
     }
+
+    fs.writeFileSync(
+      path.join(temporaryInstallPath, "package.json"),
+      `{
+  "dependencies":{
+  "@dataform/core": "${workflowSettingsDataformCoreVersion}"
+  }
+}`
+    );
+
+    await promisify(exec)("npm i --ignore-scripts", {
+      cwd: temporaryInstallPath
+    });
+
+    fs.copySync(path.join(temporaryInstallPath, "node_modules"), resolvedProjectPath);
```

Comment:
wait. you are still mutating the filesystem in the project directory. we should be doing the opposite: copy the project directory into /tmp, and then compile everything there.